README.rst
Welcome to PyU4V 9.2
====================

| |Maintenance| |OpenSource| |AskUs| |License| |Test| |Build| |Docs|
| |Language| |PyVersions| |Unisphere| |Platform| |DTotal| |DMonth| |DWeek|

Overview
--------

PyU4V is a Python module that simplifies interaction with the Unisphere for
PowerMax REST API.  It wraps REST calls with simple APIs that abstract the HTTP
request and response handling.

Full documentation and user guides can be found in PyU4V's ReadTheDocs_.

Note
   You can get the Unisphere for PowerMax REST documentation by
   navigating to a URL in your local instance of Unisphere for PowerMax.
   Navigate to ``https://{ip-address}:{port}/univmax/restapi/docs``
   where ``{ip-address}`` is the IP address of your Unisphere server and
   ``{port}`` is the port it is listening on.

PyU4V Version 9.2
-----------------

+-------------------------------+----------------------------+
| **Author**                    | Dell EMC                   |
+-------------------------------+----------------------------+
| **PyU4V Version**             | 9.2.1.4                    |
+-------------------------------+----------------------------+
| **Minimum Unisphere Version** | 9.2.0.1                    |
+-------------------------------+----------------------------+
| **Array Model**               | VMAX-3, VMAX AFA, PowerMax |
+-------------------------------+----------------------------+
| **Array uCode**               | HyperMax OS, PowerMax OS   |
+-------------------------------+----------------------------+
| **Platforms**                 | Linux, Windows             |
+-------------------------------+----------------------------+
| **Python**                    | 3.6, 3.7, 3.8, 3.9         |
+-------------------------------+----------------------------+
| **Requires**                  | Requests_, Six_, urllib3_  |
+-------------------------------+----------------------------+

Note
    If you want to continue to use Unisphere 8.4.x or 9.0.x with PyU4V you will
    need to remain on PyU4V 3.1.x. There is no support for PyU4V 9.x with any
    version of Unisphere older than 9.1.x

Note
    PyU4V version 9.x is compatible with scripts written for PyU4V versions
    >= 3.x, there is **zero** support or compatibility for PyU4V 2.x or earlier
    scripts in later versions of PyU4V. If you have scripts written which
    specifically target Unisphere REST 8.4, 9.0 or 9.1 endpoints these are
    still accessible via PyU4V 9.2.x however you will need to ensure you are
    passing the version required when performing these calls as PyU4V 9.2 will
    default to using 9.2 endpoints exclusively.  You will also need to pay
    special attention to any REST JSON payloads in custom scripts as payloads
    are subject to change between major Unisphere REST releases.

Installation
------------

Note
    A full installation guide can be found in PyU4V's ReadTheDocs_ which
    includes additional configuration options around PyU4V logging.

PyU4V can be installed from source, via ``pip``, or run directly from the
source directory. To clone PyU4V from source use ``git``::

    $ git clone https://github.com/dell/PyU4V

To install from source navigate into the new ``PyU4V`` directory and use
``pip``::

    $ cd PyU4V/
    $ pip install .

Installing via ``pip`` without cloning from source can be achieved by
specifying ``PyU4V`` as the install package for ``pip``::

    $ pip install PyU4V
    # Install a specific version
    $ pip install PyU4V==9.2.1.1

Copy the sample ``PyU4V.conf`` provided with PyU4V to either your working
directory or within a directory named ``.PyU4V`` in your current users home
directory. The ``.sample`` suffix has to be removed for the configuration file
to become valid for loading by PyU4V::

    $ mkdir ~/.PyU4V
    $ cp PyU4V/PyU4V.conf.sample ~/.PyU4V/PyU4V.conf

Note
    If ``PyU4V.conf`` is present in both the current working directory and the
    current user's home directory, the version of ``PyU4V.conf`` in the current
    working directory will take precedence.

Edit PyU4V configuration settings in ``PyU4V.conf`` under the ``[setup]``
heading, these setting will need to reflect your environment configuration::

    [setup]
    username=pyu4v-user
    password=secret-pass
    server_ip=10.0.0.75
    port=8443
    array=00012345678
    verify=/path-to-file/server_hostname.pem

Alternatively, you can pass some or all of these details on initialisation.
Environment configuration values ``password``, ``username``, ``server_ip``,
``port``, and ``array`` **must** be set either in the config file or on
initialisation. SSL verification as indicated by the ``verify`` key in
``PyU4V.conf`` is discussed in the next section.

SSL CONFIGURATION
-----------------

In order to enable SSL enabled communication between your host and the
Unisphere server there are some additional steps required. First you must
extract the CA certificate from Unisphere then either add it to the system
certificate bundle or specify the path to the cert in ``PyU4V.conf``. We will
demonstrate both approaches here.

Get the CA certificate of the Unisphere server::

    $ openssl s_client -showcerts -connect {server_hostname}:8443 \
    </dev/null 2>/dev/null|openssl x509 -outform PEM > {cert_name}.pem

    # Example
    $ openssl s_client -showcerts -connect 10.0.0.75:8443 \
    </dev/null 2>/dev/null|openssl x509 -outform PEM > unisphere91.pem

Where ``{server_host_ip}`` is the hostname or IP address of your Unisphere
server and ``{cert_name}`` is the name for your CA cert. This pulls the CA cert
file from the instance of Unisphere at ``10.0.0.75:8443`` and saves it as a
``.pem`` file.

To add the cert to a CA certificate bundle, copy the ``.pem`` file to the
system certificate directory and update the CA certificate database::

    # cp {cert_name}.pem /usr/share/ca-certificates/{cert_name}.crt
    # dpkg-reconfigure ca-certificates
    # update-ca-certificates

Once the above steps are complete you will need to specify ``verify=True`` in
``PyU4V.conf`` for PyU4V to load the required Unisphere CA cert from the system
certificate bundle::

    [setup]
    verify=True

Alternatively you can skip adding the certificate to a certificate bundle and
pass it directly on PyU4V initialisation or specify the path to the certificate
directly in ``PyU4V.conf``::

    [setup]
    verify=/path/to/file/{cert_name}.pem

Initialise PyU4V Connection
---------------------------

Initialising PyU4V in your Python scripts is as simple as importing the library
and initialising the connection (assuming you have ``PyU4V.conf`` configured as
outlined in the previous section).

.. code-block:: python

    import PyU4V

    conn = PyU4V.U4VConn()
    conn.common.get_unisphere_version()
    >> {'version': 'V9.2.0.1'}

If you wish to query another array without changing the configuration file,
call the connection ``set_array_id()`` function:

.. code-block:: python

    conn.set_array_id('000197123456')

The various types of functionality provided by PyU4V is separated into logical
sections such as ``replication``, ``provisioning``, and ``performance``. For a
full API breakdown by section and some usage example please refer to the
PyU4V ReadTheDocs_.

Support, Bugs, Issues
---------------------

Please file support requests, bugs, and issues on the PyU4V GitHub-Issues_
page for this project. For further information on opening an issue and
recommended issue templates please see the PyU4V ReadTheDocs_.

For questions asked on StackOverFlow_, please tag them with ``Dell``,
``Dell EMC``, ``PowerMax``, and ``PyU4V`` to maximise the chances of the
correct community members assisting.

Contributing
------------

PyU4V is built to be used openly by everyone, and in doing so we encourage
everyone to submit anything they may deem to be an improvement, addition, bug
fix, or other change which may benefit other users of PyU4V.

There are some requirements when submitting for PyU4V, such as coding
standards, building unit tests and continuous integration tests, and going
through a formal code review process, however anyone familiar with open source
development will be familiar with this process.  There are a number of core
PyU4V reviewers and once a submission has approvals from two or more core
reviewers and all tests are running cleanly then the request will be merged
with the upstream PyU4V repo.

For a full breakdown of contribution requirements, coding standards, submitting
and everything else in between please refer to PyU4V ReadTheDocs_.

Tools
-----

Please refer to the Tools section of ReadTheDocs_ for OpenStack functionality
to migrate volumes to the new REST masking view structure.

Disclaimer
----------

Unless required by applicable law or agreed to in writing, software distributed
under the Apache 2.0 License is distributed on an "AS IS" BASIS, WITHOUT
WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
License for the specific language governing permissions and limitations under
the License.

.. BadgeLinks

.. |Maintenance| image:: https://img.shields.io/badge/Maintained-Yes-blue
   :target: https://github.com/dell/PyU4V/commits/master
.. |OpenSource| image:: https://img.shields.io/badge/Open%20Source-Yes-blue
   :target: https://github.com/dell/PyU4V
.. |AskUs| image:: https://img.shields.io/badge/Ask%20Us...-Anything-blue
   :target: https://github.com/dell/PyU4V/issues
.. |License| image:: https://img.shields.io/badge/License-Apache%202.0-blue
   :target: https://github.com/dell/PyU4V/blob/master/LICENSE
.. |Test| image:: https://img.shields.io/badge/Tests-Passing-blue
.. |Build| image:: https://img.shields.io/badge/Build-Passing-blue
.. |Docs| image:: https://img.shields.io/badge/Docs-Passing-blue
.. |Language| image:: https://img.shields.io/badge/Language-Python%20-blue
   :target: https://www.python.org/
.. |PyVersions| image:: https://img.shields.io/badge/Python-3.6%20%7C%203.7%20%7C%203.8%20%7C%203.9-blue
   :target: https://github.com/dell/PyU4V/blob/master/README.rst
.. |Platform| image:: https://img.shields.io/badge/Platform-Linux%20%7C%20Windows-blue
   :target: https://github.com/dell/PyU4V/blob/master/README.rst
.. |Unisphere| image:: https://img.shields.io/badge/Unisphere-9.2.0.1-blue
   :target: https://www.dell.com/support/home/us/en/19/product-support/product/unisphere-powermax/overview
.. |DTotal| image:: https://pepy.tech/badge/pyu4v
   :target: https://pepy.tech/project/pyu4v
.. |DMonth| image:: https://pepy.tech/badge/pyu4v/month
   :target: https://pepy.tech/project/pyu4v/month
.. |DWeek| image:: https://pepy.tech/badge/pyu4v/week
   :target: https://pepy.tech/project/pyu4v/week

.. README URL Links

.. _Requests: https://realpython.com/python-requests/
.. _Six: https://six.readthedocs.io/
.. _urllib3: https://urllib3.readthedocs.io/en/latest/
.. _ReadTheDocs: https://pyu4v.readthedocs.io/en/latest/
.. _GitHub-Issues: https://github.com/dell/PyU4V/issues
.. _StackOverFlow: https://stackoverflow.com/search?q=PyU4V

ChangeLog
PyU4V Change Log
================
version 9.2.1.4
- added enable/disable consistency feature to modify_storage_group_srdf
  function in replication
- Updated add_existing_volume_to_storage_group to create RDF pairings.
- made srdfg number mandatory in delete_storage_group_srdf, added filters
  options, removed unit test test_delete_storagegroup_srdf_no_number
- updated constants file ISCSI_IQN_SEARCH_PATTERN = (
    '^iqn.\\d{1,4}-\\d{2}.(?:com|org).\\w{1,}:(?:\\d{2}:)*'
    '[0-9a-fA-F]{1,32}$||.[0-9A-Fa-f]{16}')
    added test to ci_utils.py test_iscsi_regex
- added relink to modify_storage_group_snapshot and
  modify_storage_group_snapshot_snapsetid


version 9.2.1.3
- added enable_mobility id to provisioning functions create_storage_group,
  add_new_volume_to_storage_group, create_volume_from_storage_group_return_id
- added get_snapshot_policy_storage_group_list to snapshot_policy functions

Version 9.2.1.2
- support for changing local users passwords via REST, requires Unipshere 9.2.1
  new function change_local_user_password

=======

Version 9.2.1.1 - released 15/02/21
- Support for Snapshot Policy compliance
- Refactoring of snapshot policy modification by operation.

Version 9.2.0.8 - released 05/02/21
- Update __init__.py to reflect the correct version
- Update the docs to reflect new version

Version 9.2.0.7 - released 03/02/21
===================================
- Fix for Python 3.9.1
- Fix for exception in get_storage_group_snapshot_list if no snapshot exists.
- CI fix of test_is_volume_in_replication_session.
- Upgrade of pep8.
- Addition of .gitattributes for unix LF checkin only.
- New default parameter for get metro DR details.

Version 9.2.0.5 - released 18/12/20
===================================
- Fix for system CI test to select only FE/SE/BE/RDF director:
    - PyU4V.tests.ci_tests.test_pyu4v_ci_system.CITestSystem.test_get_director_port_list
    - New optional parameter snapshot_policy_ids in create_non_empty_storage_group
      and create_empty_storage_group


Version 9.2.0.4 - released 9/12/20
==================================
- Fix for metro_dr delete function
- Fix for metro_dr added recover option
- Fix for CI tests:
    - PyU4V.tests.ci_tests.test_pyu4v_ci_system.CITestSystem._test_get_health_check_details
    - PyU4V.tests.ci_tests.test_pyu4v_ci_system.CITestSystem.test_get_ip_interface_list
    - PyU4V.tests.ci_tests.test_pyu4v_ci_system.CITestSystem.test_get_ip_interface
- Added Force Option to metro dr restore
- Application-Type header is now set by default to 'PyU4V-{version}'


Version 9.2.0.3 - released 7/12/20
==================================
- Masking View performance support has been added to performance.py
    - PyU4V.performance.PerformanceFunctions.get_masking_view_keys
    - PyU4V.performance.PerformanceFunctions.get_masking_view_stats
- Director list, Port list, IP interface list, and get IP interface details
  have been added to system.py
    - PyU4V.system.SystemFunctions.get_director_list
    - PyU4V.system.SystemFunctions.get_director_port_list
    - PyU4V.system.SystemFunctions.get_ip_interface_list
    - PyU4V.system.SystemFunctions.get_ip_interface
- Array storage level details are now available via provisioning.py
    - PyU4V.provisioning.ProvisioningFunctions.get_array
- Array level WLP capabilities are now accessible in workload_planner.py
    - PyU4V.workload_planner.WLPFunctions.get_capabilities
- array_id is now an optional parameter for the following functions:
    - PyU4V.replication.ReplicationFunctions.get_array_replication_capabilities
    - PyU4V.migration.MigrationFunctions.get_migration_info
    - PyU4V.migration.MigrationFunctions.get_array_migration_capabilities


Version 9.2.0.2 - released 14/10/20
===================================
- When generating threshold CSV files from
  PerformanceFunctions.generate_threshold_settings_csv it is now possible to
  set category as an input value to limit the CSV threshold file to only that
  performance category.
- CI tests for thresholds have been updated to target a default threshold
  marked as KPI to ensure it has both first and second threshold values
  populated before the test runs.
- Removed trailing white space from two performance metrics in the metric map,
  'BEDisk ReadResponseTime ' is now 'BEDiskReadResponseTime' in diagnostic
  category 'DeviceGroup', and, 'BEReadRequestTime ' is now 'BEReadRequestTime'
  in category 'StorageContainer'.


Version 9.2.0.1 - released 12/10/20
===================================
- Tighten base CI tests around RDF group creation and deletion.
- Update Unisphere minimum version to reflect 9.2 GA Unisphere version.


Version 9.2.0.0 - released 29/09/20
===================================

Snap id Support
---------------
Functions added:
    - get_storage_group_snapshot_snap_id_list
    - get_snapshot_snap_id_details
    - find_expired_snapvx_snapshots_by_snap_ids
    - modify_storage_group_snapshot_by_snap_id
    - restore_snapshot_by_snap_id
    - rename_snapshot_by_snap_id
    - link_snapshot_by_snap_id
    - unlink_snapshot_by_snap_id
    - delete_storage_group_snapshot_by_snap_id

Snapshot Policy Functions
-------------------------
Functions added:
    - get_snapshot_policy_list
    - get_snapshot_policy
    - create_snapshot_policy
    - modify_snapshot_policy
    - delete_snapshot_policy
Functions modified:
    - create_storage_group

System Enhancements
-------------------
- Users can now export (download) and import (upload) both Unisphere and array
  system settings combined or individually using system calls.  Settings can
  be written to a zip file or returned to the caller function as binary data
  so file writing can be handled outside of PyU4V or uploaded directly to
  Unisphere without the need for file handling. Users can call the following
  system functions for settings download/upload:
    - download_all_settings()
    - download_unisphere_settings()
    - download_system_settings()
    - upload_settings()

PyU4V/__init__.py
# -*- coding: utf-8 -*-
# __________         ____ ___  _________   ____
# \______   \___.__.|    |   \/  |  \   \ /   /
#  |     ___<   |  ||    |   /   |  |\   Y   /
#  |    |    \___  ||    |  /    ^   /\     /
#  |____|    / ____||______/\____   |  \___/
#            \/                  |__|
# Copyright (c) 2020 Dell Inc. or its subsidiaries.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""__init__.py."""

from .univmax_conn import U4VConn  # noqa: F401

__title__ = 'pyu4v'
__version__ = '9.2.1.4'
__author__ = 'Dell EMC or its subsidiaries'
__license__ = 'Apache 2.0'
__copyright__ = 'Copyright 2020 Dell EMC Inc'


PyU4V/utils/constants.py
# Copyright (c) 2020 Dell Inc. or its subsidiaries.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""constants.py."""

# Configuration constants
SETUP = 'setup'
ARRAY = 'array'
R_ARRAY = 'remote_array'
R_ARRAY_2 = 'remote_array_2'
USERNAME = 'username'
PASSWORD = 'password'
SERVER_IP = 'server_ip'
PORT = 'port'
VERIFY = 'verify'

# HTTP constants
GET = 'GET'
POST = 'POST'
PUT = 'PUT'
DELETE = 'DELETE'
CONTENT_TYPE = 'content-type'
ACCEPT = 'accept'
ACCEPT_ENC = 'Accept-Encoding'
USER_AGENT = 'user-agent'
APP_TYPE = 'application-type'
APP_JSON = 'application/json'
APP_OCT = 'application/octet-stream'
APP_MPART = 'multipart/form-data'

# Unisphere REST URI constants
PYU4V_VERSION = '9.2.1.4'
UNISPHERE_VERSION = '92'
VERSION = 'version'
ITERATOR = 'Iterator'
PAGE = 'page'
JOB = 'job'
SYMMETRIX = 'symmetrix'
SLOPROVISIONING = 'sloprovisioning'
PROVISIONING = 'provisioning'
STORAGEGROUP = 'storagegroup'
REPLICATION = 'replication'
METRO_DR = 'metrodr'
WLP = 'wlp'
MIGRATION = 'migration'
DSA = 'dsa'
SYSTEM = 'system'
LOCAL_USER = 'local_user'
VOLUME = 'volume'
VVOL = 'vvol'
COMMON = 'common'
HEADROOM = 'headroom'
ENVIRONMENT = 'environment'
CAPABILITIES = 'capabilities'
DIRECTOR = 'director'
PORTGROUP = 'portgroup'
HOST = 'host'
HOSTGROUP = 'hostgroup'
INITIATOR = 'initiator'
MASKINGVIEW = 'maskingview'
CONNECTIONS = 'connections'
SLO = 'slo'
WORKLOADTYPE = 'workloadtype'
SRP = 'srp'
COMPRESSIBILITY_REPORT = 'compressibility_report'
SG_DEMAND_REPORT = 'storage_group_demand_report'
SNAPSHOT = 'snapshot'
GENERATION = 'generation'
SNAP_ID = 'snapid'
RDFG = 'rdf_group'
RDF_GROUP = 'rdf_group'
GB_HEADROOM = 'gbHeadroom'
RDF_DIRECTOR = 'rdf_director'
REMOTE_PORT = 'remote_port'
ALERT = 'alert'
ALERT_SUMMARY = 'alert_summary'
COMPLIANCE = 'compliance'

# Status Codes
STATUS_200 = 200
STATUS_201 = 201
STATUS_202 = 202
STATUS_204 = 204
STATUS_401 = 401
STATUS_404 = 404

# Job constants
INCOMPLETE_LIST = ['created', 'scheduled', 'running',
                   'validating', 'validated']
CREATED = 'created'
SUCCEEDED = 'succeeded'
ASYNCHRONOUS = 'ASYNCHRONOUS'
ASYNC_UPDATE = {'executionOption': ASYNCHRONOUS}
CREATE_VOL_STRING = 'Creating new Volumes'

# Replication Modes
ASYNCHRONOUS_CC = 'Asynchronous'
ADAPTIVE_COPY = 'AdaptiveCopyDisk'
DISABLECONSISTENCY = 'DisableConsistency'
ENABLECONSISTENCY = 'EnableConsistency'
ESTABLISH = 'Establish'
FAILBACK = 'Failback'
FAILOVER = 'Failover'
RECOVER = 'Recover'
RESTORE = 'Restore'
RESUME = 'Resume'
SETBIAS = 'SetBias'
SETMODE = 'SetMode'
SPLIT = 'Split'
SUSPEND = 'Suspend'
SWAP = 'Swap'
UPDATE_R1 = 'UpdateR1'
# RDFG Constants

RDFG_ACTIONS = {
    'MOVE': 'Move', 'ADD_PORTS': 'add_ports',
    'REMOVE_PORTS': 'remove_ports', 'SET_LABEL': 'set_label'}
METRO_DR_ACTIONS = {
    'ESTABLISH': ESTABLISH, 'SPLIT': SPLIT, 'SUSPEND': SUSPEND,
    'RECOVER': RECOVER, 'RESTORE': RESTORE, 'FAILOVER': FAILOVER,
    'FAILBACK': FAILBACK, 'SETMODE': SETMODE, 'UPDATER1': UPDATE_R1
}
METRO_DR_ACTION_PARAMS = {
    'ESTABLISH': 'establish', 'SPLIT': 'split', 'SUSPEND': 'suspend',
    'RESTORE': 'restore', 'FAILOVER': 'failover',
    'FAILBACK': 'failback', 'SETMODE': 'set_mode', 'UPDATER1': 'update_r1'
}

METRO_DR_REPLICATION_MODE = {'ADAPTIVECOPYDISK': 'AdaptiveCopyDisk',
                             'ASYNCHRONOUS': 'Asynchronous'}

# Director constants
DIRECTOR_ID = 'directorId'
PORT_ID = 'portId'
SYMMETRIX_PORT_KEY = 'symmetrixPortKey'
NUM_OF_PORTS = 'num_of_ports'
NUM_OF_CORES = 'num_of_cores'
IP_INTERFACE = 'ipinterface'
IP_INTERFACE_ID = 'ipInterfaceId'

# Host constants
HOST_ID = 'hostId'
ENABLED_FLAGS = 'enabled_flags'
CONSISTENT_LUN = 'consistent_lun'
NUM_OF_INITIATORS = 'num_of_initiators'
HOST_LUN_ADDRESS = 'host_lun_address'
OVERRIDE = 'override'
ENABLED = 'enabled'
TRUE = 'true'
FALSE = 'false'
SPC2_PROTOCOL_VERSION = 'spc2_protocol_version'
SPC2_PROTOCOL_VERSION_FLAG = 'SPC2_Protocol_Version(SPC2)'
SCSI_3 = 'scsi_3'
SCSI_3_FLAG = 'SCSI_3(SC3)'
OPENVMS = 'openvms'
OPENVMS_FLAG = 'OpenVMS(OVMS)'
AVOID_RESET_BROADCAST = 'avoid_reset_broadcast'
AVOID_RESET_BROADCAST_FLAG = 'Avoid_Reset_Broadcast(ARB)'
VOLUME_SET_ADDRESSING = 'volume_set_addressing'
HOST_FLAGS = {
    'disable_q_reset_on_ua': {
        OVERRIDE: FALSE, ENABLED: FALSE},
    'environ_set': {
        OVERRIDE: FALSE, ENABLED: FALSE},
    VOLUME_SET_ADDRESSING: {
        OVERRIDE: FALSE, ENABLED: FALSE},
    'spc2_protocol_version': {
        OVERRIDE: FALSE, ENABLED: FALSE},
    'scsi_support1': {
        OVERRIDE: FALSE, ENABLED: FALSE},
    OPENVMS: {
        OVERRIDE: FALSE, ENABLED: FALSE},
    AVOID_RESET_BROADCAST: {
        OVERRIDE: FALSE, ENABLED: FALSE},
    SCSI_3: {
        OVERRIDE: TRUE, ENABLED: TRUE},
    CONSISTENT_LUN: FALSE}

# Host Group constants
NUM_OF_HOSTS = 'num_of_hosts'
HOST_GROUP = 'host_group'
HOST_GROUP_ID = 'hostGroupId'

# SRP constants
SRP_CAPACITY = 'srp_capacity'
SRP_EFFICIENCY = 'srp_efficiency'
STORAGE_GROUP_DEMAND = 'storage_group_demand'
SRP_ID = 'srpId'
NUM_OF_DISK_GROUPS = 'num_of_disk_groups'
RESERVED_CAP_PERCENT = 'reserved_cap_percent'
TOTAL_SRDF_DSE_ALLOCATED_CAP_GB = 'total_srdf_dse_allocated_cap_gb'
RDFA_DSE = 'rdfa_dse'
DISK_GROUP_ID = 'diskGroupId'

# Storage Group constants
STORAGE_GROUP_ID_CAMEL = 'storageGroupId'
STORAGE_GROUP_ID = 'storage_group_id'
NUM_OF_CHILD_SGS = 'num_of_child_sgs'
SERVICE_LEVEL = 'service_level'
BASE_SLO_NAME = 'base_slo_name'
SLO_COMPLIANCE = 'slo_compliance'
NUM_OF_VOLS = 'num_of_vols'
NUM_OF_PARENT_SGS = 'num_of_parent_sgs'
NUM_OF_MASKING_VIEWS = 'num_of_masking_views'
NUM_OF_SNAPSHOTS = 'num_of_snapshots'
DEVICE_EMULATION = 'device_emulation'
UNPROTECTED = 'unprotected'
COMPRESSION = 'compression'

# Port Group constants
PORT_GROUP = 'port_group'
PORT_GROUP_ID = 'portGroupId'

# Masking View constants
MASKING_VIEW_ID = 'maskingViewId'

# Volume constants
VOLUME_ID_CAMEL = 'volumeId'
VOLUME_ID = 'volume_id'
DEVICE_ID = 'device_id'
CAP_GB = 'cap_gb'
ALLOCATED_PERCENT = 'allocated_percent'
VOLUME_IDENTIFIER = 'volume_identifier'
VOLUME_NAME = 'volume_name'
WWN = 'wwn'
EFFECTIVE_WWN = 'effective_wwn'
HAS_EFFECTIVE_WWN = 'has_effective_wwn'
EMULATION = 'emulation'
TYPE = 'type'
SSID = 'ssid'
CAP_MB = 'cap_mb'
CAP_CYL = 'cap_cyl'
STATUS = 'status'
RESERVED = 'reserved'
PINNED = 'pinned'
ENCAPSULATED = 'encapsulated'
NUM_OF_FRONT_END_PATHS = 'num_of_front_end_paths'
SNAPVX_SOURCE = 'snapvx_source'
SNAPVX_TARGET = 'snapvx_target'
MOBILITY_ID_ENABLED = 'mobility_id_enabled'

# SLO constants
SLO_ID = 'sloId'
NUM_OF_STORAGE_GROUPS = 'num_of_storage_groups'

# Initiator constants
INITIATOR_ID = 'initiatorId'
LOGGED_IN = 'logged_in'
ALIAS = 'alias'
ON_FABRIC = 'on_fabric'
FCID = 'fcid'

# Regex search pattern constants
DIRECTOR_SEARCH_PATTERN = '\\w{2}-\\w{2}'
PORT_SEARCH_PATTERN = '\\d{1,3}'
WWN_SEARCH_PATTERN_16 = '[0-9a-fA-F]{16}'
WWN_SEARCH_PATTERN_32 = '[0-9a-fA-F]{32}'
ISCSI_IQN_SEARCH_PATTERN = (
    '^iqn.\\d{1,4}-\\d{2}.(?:com|org).\\w{1,}:(?:\\d{2}:)*'
    '[0-9a-fA-F]{1,32}$||.[0-9A-Fa-f]{16}')
non_iscsi_initiator_pattern = '{d}:{p}:{w}'.format(
    d=DIRECTOR_SEARCH_PATTERN, p=PORT_SEARCH_PATTERN,
    w=WWN_SEARCH_PATTERN_16)
iscsi_initiator_pattern = '{d}:{p}:{pa}'.format(
    d=DIRECTOR_SEARCH_PATTERN, p=PORT_SEARCH_PATTERN,
    pa=ISCSI_IQN_SEARCH_PATTERN)
INITIATOR_SEARCH_PATTERN = '{ni}|{i}'.format(
    ni=non_iscsi_initiator_pattern, i=iscsi_initiator_pattern)

# QOS constants
HOST_IO_LIMIT = 'hostIOLimit'
MAX_MBPS = 'maxMBPS'
DISTRIBUTION_TYPE = 'DistributionType'
HOST_IO_LIMIT_IO_SEC = 'host_io_limit_io_sec'
HOST_IO_LIMIT_MB_SEC = 'host_io_limit_mb_sec'
DYNAMIC_DISTRIBUTION = 'dynamicDistribution'

# System Health / Disk Checks
HEALTH = 'health'
HEALTH_CHECK = 'health_check'
FAILED = 'failed'
DISK = 'disk'
TAG = 'tag'
DESCRIPTION = 'description'
ARRAY_ID = 'array_id'
TAG_NAME = 'tag_name'
SG_ID = 'storage_group_id'
SG_NUM = 'num_of_storage_groups'
ARRAY_NUM = 'num_of_arrays'
HEALTH_SCORE = 'health_score_metric'
HEALTH_CHECK_ID = 'health_check_id'
TEST_RES = 'testResult'
JOB_ID = 'jobId'
RES_LINK = 'resourceLink'
NAME = 'name'
DISK_IDS = 'disk_ids'
SPINDLE_ID = 'spindle_id'
VENDOR = 'vendor'
CAPACITY = 'capacity'
ALERT_SUMMARY_KEYS = ['alert_count', 'all_unacknowledged_count',
                      'fatal_unacknowledged_count',
                      'critical_unacknowledged_count',
                      'warning_unacknowledged_count',
                      'info_unacknowledged_count',
                      'minor_unacknowledged_count',
                      'normal_unacknowledged_count',
                      'all_acknowledged_count', 'fatal_acknowledged_count',
                      'critical_acknowledged_count',
                      'warning_acknowledged_count',
                      'info_acknowledged_count', 'minor_acknowledged_count',
                      'normal_acknowledged_count']
SNAPSHOT_POLICY_INTERVALS = ['10 Minutes', '12 Minutes', '15 Minutes',
                             '20 Minutes', '30 Minutes', '1 Hour',
                             '2 Hours', '3 Hours', '4 Hours', '6 Hours',
                             '8 Hours', '12 Hours', '1 Day', '7 Days']
ASSOCIATE_TO_STORAGE_GROUPS = 'AssociateToStorageGroups'
DISASSOCIATE_FROM_STORAGE_GROUPS = 'DisassociateFromStorageGroups'
SNAPSHOT_POLICY = 'snapshot_policy'
SNAPSHOT_POLICY_NAME_FOR_TEST = 'DailyDefault'
MODIFY_POLICY = 'Modify'
SUSPEND_POLICY = 'Suspend'
RESUME_POLICY = 'Resume'
SNAPSHOT_POLICY_ACTIONS = [MODIFY_POLICY, SUSPEND_POLICY, RESUME_POLICY,
                           ASSOCIATE_TO_STORAGE_GROUPS,
                           DISASSOCIATE_FROM_STORAGE_GROUPS]

# Import/Export Settings
SETTINGS_FILENAME_TEMPLATE = 'PyU4V-SettingsExport'
SETTINGS = 'settings'
EXPORT_FILE = 'exportfile'
IMPORT_FILE = 'importfile'
SRC_ARRAY = 'source_array'
TGT_ARRAYS = 'target_arrays'
FILE_PASSWORD = 'file_password'
ZIP_FILE = 'zip_file'
ZIP_SUFFIX = '.zip'
PDF_SUFFIX = '.pdf'
FILE_READ_MODE = 'rb'
FILE_WRITE_MODE = 'wb'
ALL_SETTINGS = 'all'
EXCLUDE_UNI_SETTINGS = 'exclude_unisphere_setting_options'
EXCLUDE_SYS_SETTINGS = 'exclude_system_setting_options'
UNI_ALERT_SETTINGS = 'alert_notification_settings'
UNI_PERF_PREF_SETTINGS = 'performance_preference_settings'
UNI_PERF_USER_SETTINGS = 'performance_user_templates'
UNI_PERF_METRIC_SETTINGS = 'performance_metric_settings'
SYS_ALERT_SETTINGS = 'alert_policy_settings'
SYS_ALERT_NOTIFI_SETTINGS = 'alert_level_notification_settings'
SYS_THRESH_SETTINGS = 'system_threshold_settings'
SYS_PERF_THRESH_SETTINGS = 'performance_threshold_settings'

# Audit Log Constants
AUDIT_LOG_FILENAME_TEMPLATE = 'PyU4V-AuditLogRecord'
AUDIT_LOG_FILENAME = 'auditlogfilename'
AUDIT_RECORD_PATH = 'audit_record_path'
AUDIT_LOG_RECORD = 'audit_log_record'
AUDIT_RECORD_TIME = 'audit_record_time'
BINARY_DATA = 'binary_data'
RECORD_ID = 'record_id'
HOST_NAME = 'hostname'
CLIENT_HOST = 'client_host'
MESSAGE = 'message'
ACTIVITY_ID = 'activity_id'
APP_ID = 'application_id'
APP_VERSION = 'application_version'
TASK_ID = 'task_id'
PROCESS_ID = 'process_id'
VENDOR_ID = 'vendor_id'
OS_TYPE = 'os_type'
OS_REV = 'os_revision'
API_LIB = 'api_library'
API_VER = 'api_version'
AUDIT_CLASS = 'audit_class'
ACTION_CODE = 'action_code'
FUNC_CLASS = 'function_class'
SUCCESS = 'success'
COUNT = 'count'

# Date/Time
STR_TIME_FORMAT = '%Y%m%d%H%M%S'
